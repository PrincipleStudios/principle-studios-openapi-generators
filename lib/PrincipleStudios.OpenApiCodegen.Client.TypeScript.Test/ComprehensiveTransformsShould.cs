using PrincipleStudios.OpenApi.Transformations;
using PrincipleStudios.OpenApi.TypeScript;
using Snapshooter.Xunit;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using static PrincipleStudios.OpenApiCodegen.TestUtils.DocumentHelpers;

namespace PrincipleStudios.OpenApiCodegen.Client.TypeScript
{
    using static OptionsHelpers;
    public class ComprehensiveTransformsShould
    {
        [Trait("Category", "Integration")]
        [MemberData(nameof(ValidFileNames))]
        [Theory]
        public void CoverFullFiles(string name)
        {
            var document = GetDocument(name);
            var options = LoadOptions();

            var transformer = document.BuildTypeScriptOperationSourceProvider("", options);
            OpenApiTransformDiagnostic diagnostic = new();

            var entries = transformer.GetSources(diagnostic).ToArray();

            var sb = new StringBuilder();
            foreach (var entry in entries)
            {
                sb.AppendLine($"exports[`{entry.Key}`] = `{entry.SourceText}`;");
                sb.AppendLine();
            }
            Snapshot.Match(sb.ToString(), $"{nameof(ComprehensiveTransformsShould)}.{TypeScriptNaming.ToTitleCaseIdentifier(name, options.ReservedIdentifiers())}");
            Assert.Empty(diagnostic.Errors);
        }

        [Trait("Category", "Integration")]
        [MemberData(nameof(InvalidFileNames))]
        [Theory]
        public void ReportDiagnosticsForMissingReferences(string name)
        {
            var document = GetDocument(name);
            var options = LoadOptions();

            var transformer = document.BuildTypeScriptOperationSourceProvider("", options);
            OpenApiTransformDiagnostic diagnostic = new();

            var entries = transformer.GetSources(diagnostic).ToArray();

            Snapshot.Match(diagnostic.Errors.Select(err => new { Context = err.Context.ToOpenApiPathContextString(), Message = err.Message }).ToArray(), $"Diagnostics.{TypeScriptNaming.ToTitleCaseIdentifier(name, options.ReservedIdentifiers())}");
        }

        public static IEnumerable<object[]> ValidFileNames =>
            from fileIndex in GetValidDocumentIndices()
            select new object[] { GetDocumentName(fileIndex) };

        public static IEnumerable<object[]> InvalidFileNames =>
            from fileIndex in GetInvalidDocumentIndices()
            select new object[] { GetDocumentName(fileIndex) };

    }
}
