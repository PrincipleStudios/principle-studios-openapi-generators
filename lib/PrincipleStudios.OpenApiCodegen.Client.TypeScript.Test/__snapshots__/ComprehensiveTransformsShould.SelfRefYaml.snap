exports[`operations/getData.ts`] = `/*
 * Self-ref
 *
 * A sample API that demonstrates a self-referencing object
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */
import { HttpHeaders, HttpQuery, RequestOpts, ResponseArgs, throwIfNullOrUndefined, encodeURI, RequestConversion, StandardResponse, TransformRequest } from '@principlestudios/openapi-codegen-typescript';
import {
    Target,
} from './../models/Target';

export type RequestParams = {
};
export type RequestBodies = {
};

export type StructuredResponses = {
    /** Gets any kind of JSON data */
    200: {
        'application/json': Target;
    };
};
export type Responses = 
    | StandardResponse<200, 'application/json', Target>;

export function getData(params: RequestParams): RequestOpts;
export function getData(params: RequestParams): RequestOpts {
    throwIfNullOrUndefined(params, 'params');

    const headers: HttpHeaders = {};

    return {
        path: '/data',
        method: 'POST',
        headers,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            200,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.getResponseHeader("Content-Type"),
        data: res.response,
        response: res,
    }) as Responses;

export const conversion: RequestConversion<RequestParams, {}, Responses, 'no-body'> = {
    request: getData,
    response: responseConversion,
};

`;

exports[`operations/index.ts`] = `/*
 * Self-ref
 *
 * A sample API that demonstrates a self-referencing object
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */
import { conversion as getData } from './getData';

export default {
    getData,
} as const;
`;

exports[`models/Target.ts`] = `/*
 * Self-ref
 *
 * A sample API that demonstrates a self-referencing object
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */

/**
 * @export
 * @interface Target
 */
export type Target = {
    /**
     * @type Target | null
     */
    target?: Target | null;
    /**
     * @type number
     */
    id: number;
};
`;

